File: C++/代码随想录/Binary_Tree/Traversal mode of Binary tree/levelOrder/515. Find Largest Value in Each Tree Row.py
from collections import deque
from typing import List
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def largestValues(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []
        res = []
        queue = deque([root])
        while queue:
            level_max = float('-inf')
            for _ in range(len(queue)):
                node = queue.popleft()
                level_max = max(level_max, node.val)
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            res.append(level_max)
        return res

# Test case
root = TreeNode(1, TreeNode(3, TreeNode(5), TreeNode(3)), TreeNode(2, None, TreeNode(9)))
sol = Solution()
print(sol.largestValues(root))  # [1, 3, 9]

File: C++/代码随想录/Binary_Tree/Traversal mode of Binary tree/levelOrder/515. Find Largest Value in Each Tree Row.cpp
// 515. Find Largest Value in Each Tree Row
// https://leetcode.com/problems/find-largest-value-in-each-tree-row/
// clang++ 515.\ Find\ Largest\ Value\ in\ Each\ Tree\ Row.cpp -std=c++17

#include "../../../printUtils.hpp"
#include <limits>
#include <vector>

// 找出每一行中的最大值
template <typename T>
std::vector<T> largestValues(TreeNode<T>* root) {
    std::vector<T> result;
    if (root == nullptr) return result;

    std::queue<TreeNode<T>*> q;
    q.push(root);

    while (!q.empty()) {
        int size = q.size();
        T maxVal = std::numeric_limits<T>::lowest(); 
        for (int i = 0; i < size; i++) {
            TreeNode<T>* curr = q.front();
            q.pop();

            maxVal = std::max(maxVal, curr->val);

            if (curr->left) q.push(curr->left);
            if (curr->right) q.push(curr->right);
        }
        result.push_back(maxVal);
    }
    return result;
}

int main() {
    TreeNode<int>* root = buildTree({1, 3, 2, 5, 3, -1, 9}, -1);
    std::cout << "Largest value in each tree row: ";
    printArray(largestValues(root));
    deleteTree(root);
    
    return 0;
}
